gcs/gcs.go

```diff
@@ -2,219 +2,143 @@ package gcs
+// NewGcsFileWriter will create a new GCS file reader.
+func NewGcsFileReader(ctx context.Context, projectID, bucketName, name string) (*File, error) {
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create storage client: %w", err)
 	}
-	return res.Open(name)
-}
 
-func NewGcsFileReaderWithClient(ctx context.Context, client *storage.Client, projectId string, bucketName string, name string) (source.ParquetFile, error) {
-	res := &GcsFile{
-		ProjectId:      projectId,
-		BucketName:     bucketName,
-		Ctx:            ctx,
-		Client:         client,
-		externalClient: true,
-		FilePath:       name,
+	r, err := NewGcsFileReaderWithClient(ctx, client, projectID, bucketName, name)
+	if err != nil {
+		return nil, err
 	}
-	return res.Open(name)
+
+	// Set externalClient to false so we close it when calling `Close`.
+	r.externalClient = false
+
+	return r, nil
 }
 
-func (self *GcsFile) Create(name string) (source.ParquetFile, error) {
-	var err error
-	gcs := new(GcsFile)
-	if self.Client == nil {
-		gcs.Client, err = storage.NewClient(self.Ctx)
-		gcs.externalClient = false
-	} else {
-		gcs.Client = self.Client
-		gcs.externalClient = self.externalClient
-	}
-	gcs.FilePath = name
+// NewGcsFileWriter will create a new GCS file reader with the passed client.
+func NewGcsFileReaderWithClient(ctx context.Context, client *storage.Client, projectID, bucketName, name string) (*File, error) {
+	obj := client.Bucket(bucketName).Object(name)
+
+	reader, err := gcsobj.NewReader(ctx, obj)
 	if err != nil {
-		return gcs, err
+		return nil, fmt.Errorf("failed to create new reader: %w", err)
 	}
-	// must use existing bucket
-	gcs.Bucket = gcs.Client.Bucket(self.BucketName)
-	obj := gcs.Bucket.Object(name)
-	gcs.FileWriter = obj.NewWriter(self.Ctx)
-	return gcs, err
+
+	return &File{
+		ProjectID:      projectID,
+		BucketName:     bucketName,
+		FilePath:       name,
+		gcsClient:      client,
+		gcsReader:      reader,
+		object:         obj,
+		ctx:            ctx,
+		externalClient: true,
+	}, nil
 }
 
-func (self *GcsFile) Open(name string) (source.ParquetFile, error) {
-	var err error
-	gcs := new(GcsFile)
-	if self.Client == nil {
-		gcs.Client, err = storage.NewClient(self.Ctx)
-		gcs.externalClient = false
-	} else {
-		gcs.Client = self.Client
-		gcs.externalClient = self.externalClient
-	}
-	if err != nil {
-		return gcs, err
-	}
+// Open will create a new GCS file reader/writer and open the object named as
+// the passed named. If name is left empty the same object as currently opened
+// will be re-opened.
+func (g *File) Open(name string) (source.ParquetFile, error) {
 	if name == "" {
-		gcs.FilePath = self.FilePath
-	} else {
-		gcs.FilePath = name
+		name = g.FilePath
 	}
-	// must use existing bucket
-	gcs.Bucket = gcs.Client.Bucket(self.BucketName)
-	obj := gcs.Bucket.Object(gcs.FilePath)
-	attrs, err := obj.Attrs(self.Ctx)
-	if err != nil {
-		return gcs, err
+
+	if !externalClient {
```

Comment:
I first just checked if you had an existing client or not, but that would not work if you do:

```go
f1, _ := NewGcsFileReader(...)
f2, _ := f1.Open("")
f1.Close() // Would close client f2 is using
```

But reading the code in `parquet-go` some more I realise we use one reader per column which could potentially be a very huge number of clients.

WDYT is the right approach here? Should I just re-use the client if it already exists no matter where it's coming from and assume that no one is is calling `Close` before done with all readers? I guess you are the only consumer of this package in `parquet-go` anyway and i that code only calls close on `ReadStop` when reading is done.